File: tezos/ac-login.mjs
#!/usr/bin/env node
/**
 * ac-login - CLI authentication for Aesthetic Computer
 * 
 * Uses OAuth 2.0 Authorization Code Flow with PKCE + local callback server
 * Opens browser → user logs in → callback returns to localhost → token saved
 * 
 * Usage: 
 *   node ac-login.mjs         - Login (opens browser)
 *   node ac-login.mjs status  - Check login status
 *   node ac-login.mjs logout  - Clear stored token
 *   node ac-login.mjs token   - Print access token (for scripts)
 */

import { promises as fs } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';
import { exec } from 'child_process';
import http from 'http';
import crypto from 'crypto';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Config - matches vscode-extension and kidlisp.com
const AUTH0_DOMAIN = 'hi.aesthetic.computer';
const AUTH0_CLIENT_ID = 'LVdZaMbyXctkGfZDnpzDATB5nR0ZhmMt';
const CALLBACK_PORT = 44233;  // Matches Auth0 allowed callback URL
const REDIRECT_URI = `http://localhost:${CALLBACK_PORT}/callback`;
const TOKEN_FILE = join(process.env.HOME, '.ac-token');

// PKCE helpers
function base64URLEncode(buffer) {
  return buffer.toString('base64')
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=/g, '');
}

function generateCodeVerifier() {
  return base64URLEncode(crypto.randomBytes(32));
}

function generateCodeChallenge(verifier) {
  const hash = crypto.createHash('sha256').update(verifier).digest();
  return base64URLEncode(hash);
}

// Open browser (devcontainer-aware)
function openBrowser(url) {
  const browserCmd = process.env.BROWSER;
  if (browserCmd) {
    exec(`${browserCmd} "${url}"`, (err) => {
      if (err) console.error('Failed to open browser:', err.message);
    });
  } else {
    const cmd = process.platform === 'darwin' ? 'open' :
                process.platform === 'win32' ? 'start' : 'xdg-open';
    exec(`${cmd} "${url}"`, (err) => {
      if (err) console.error('Failed to open browser:', err.message);
    });
  }
}

// Main login flow
async function login() {
  console.log('╔══════════════════════════════════════════════════════════════╗');
  console.log('║  🔐 Aesthetic Computer CLI Login                             ║');
  console.log('╚══════════════════════════════════════════════════════════════╝\n');
  
  const codeVerifier = generateCodeVerifier();
  const codeChallenge = generateCodeChallenge(codeVerifier);
  const state = crypto.randomBytes(16).toString('hex');
  
  try {
    const { tokens, user } = await startLocalCallbackServer(state, codeVerifier, codeChallenge);
    
    const displayName = user.handle ? `@${user.handle}` : user.email || user.name;
    console.log('\n✅ Authentication successful!\n');
    console.log(`👤 User: ${displayName}`);
    console.log(`💾 Token stored in: ${TOKEN_FILE}`);
    console.log('\nYou can now use authenticated API calls.\n');
    
    return tokens;
  } catch (err) {
    console.error('\n❌ Login failed:', err.message);
    process.exit(1);
  }
}

// Start local callback server
async function startLocalCallbackServer(state, codeVerifier, codeChallenge) {
  return new Promise((resolve, reject) => {
    const server = http.createServer(async (req, res) => {
      const url = new URL(req.url, `http://localhost:${CALLBACK_PORT}`);
      
      if (!url.pathname.includes('callback')) {
        res.writeHead(404);
        res.end('Not found');
        return;
      }
      
      const code = url.searchParams.get('code');
      const error = url.searchParams.get('error');
      
      if (error) {
        res.writeHead(200, { 'Content-Type': 'text/html; charset=utf-8' });
        res.end(`<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Login Failed · Aesthetic Computer</title>
  <style>
    @media (prefers-color-scheme: dark) {
      body { background: rgb(50, 30, 60); color: rgb(220, 30, 100); }
      code { color: pink; }
      mark { background-color: maroon; color: pink; }
    }
    @media (prefers-color-scheme: light) {
      body { background: rgb(255, 240, 245); color: rgb(180, 20, 80); }
      code { color: rgb(150, 50, 100); }
      mark { background-color: rgb(255, 200, 220); color: rgb(150, 20, 60); }
    }
    body {
      font-family: monospace;
      display: flex;
      width: 100vw;
      height: 100vh;
      margin: 0;
    }
    #wrapper { margin: auto; text-align: center; padding: 2em; }
    h1 { font-family: sans-serif; font-weight: normal; }
    code { font-size: 120%; opacity: 0.75; }
  </style>
</head>
<body>
  <div id="wrapper">
    <h1>❌ Authentication Failed</h1>
    <p><mark>${error}</mark></p>
    <p>${url.searchParams.get('error_description') || ''}</p>
    <p><code>You may close this page.</code></p>
  </div>
</body>
</html>`);
        server.close();
        reject(new Error(error));
        return;
      }
      
      if (!code) {
        res.writeHead(400);
        res.end('Missing authorization code');
        return;
      }
      
      try {
        const tokenResponse = await fetch(`https://${AUTH0_DOMAIN}/oauth/token`, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            grant_type: 'authorization_code',
            client_id: AUTH0_CLIENT_ID,
            code_verifier: codeVerifier,
            code: code,
            redirect_uri: REDIRECT_URI,
          }),
        });
        
        if (!tokenResponse.ok) {
          const errText = await tokenResponse.text();
          throw new Error(`Token exchange failed: ${errText}`);
        }
        
        const tokens = await tokenResponse.json();
        
        const userResponse = await fetch(`https://${AUTH0_DOMAIN}/userinfo`, {
          headers: { 'Authorization': `Bearer ${tokens.access_token}` },
        });
        
        const user = await userResponse.json();
        
        // Fetch AC handle
        let acHandle = null;
        try {
          const handleRes = await fetch(
            `https://aesthetic.computer/user?from=${encodeURIComponent(user.email)}&withHandle=true`
          );
          const handleData = await handleRes.json();
          if (handleData.handle) acHandle = handleData.handle;
        } catch (e) {}
        
        await fs.writeFile(TOKEN_FILE, JSON.stringify({
          access_token: tokens.access_token,
          refresh_token: tokens.refresh_token,
          id_token: tokens.id_token,
          expires_at: Date.now() + (tokens.expires_in * 1000),
          user: { email: user.email, name: user.name, sub: user.sub, picture: user.picture, handle: acHandle },
        }, null, 2));
        
        const displayName = acHandle ? `@${acHandle}` : user.email || user.name;
        res.writeHead(200, { 'Content-Type': 'text/html; charset=utf-8' });
        res.end(`<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Logged In · Aesthetic Computer</title>
  <style>
    @media (prefers-color-scheme: dark) {
      body { background: rgb(50, 30, 60); color: rgb(220, 30, 100); }
      code { color: pink; }
      strong { color: white; }
    }
    @media (prefers-color-scheme: light) {
      body { background: rgb(255, 240, 245); color: rgb(180, 20, 80); }
      code { color: rgb(150, 50, 100); }
      strong { color: rgb(100, 20, 60); }
    }
    body {
      font-family: monospace;
      display: flex;
      width: 100vw;
      height: 100vh;
      margin: 0;
    }
    #wrapper { margin: auto; text-align: center; padding: 2em; }
    h1 { font-family: sans-serif; font-weight: normal; }
    code { font-size: 120%; opacity: 0.75; }
  </style>
</head>
<body>
  <div id="wrapper">
    <h1>✅ Login Successful</h1>
    <p>Welcome, <strong>${displayName}</strong>!</p>
    <p><code>You may close this page.</code></p>
  </div>
  <script>setTimeout(() => window.close(), 3000)</script>
</body>
</html>`);
        
        server.close();
        resolve({ tokens, user: { ...user, handle: acHandle } });
      } catch (err) {
        res.writeHead(500, { 'Content-Type': 'text/plain' });
        res.end(`Error: ${err.message}`);
        server.close();
        reject(err);
      }
    });
    
    server.listen(CALLBACK_PORT, () => {
      const authUrl = new URL(`https://${AUTH0_DOMAIN}/authorize`);
      authUrl.searchParams.set('response_type', 'code');
      authUrl.searchParams.set('client_id', AUTH0_CLIENT_ID);
      authUrl.searchParams.set('redirect_uri', REDIRECT_URI);
      authUrl.searchParams.set('scope', 'openid profile email offline_access');
      authUrl.searchParams.set('state', state);
      authUrl.searchParams.set('code_challenge', codeChallenge);
      authUrl.searchParams.set('code_challenge_method', 'S256');
      
      console.log('🌐 Opening browser for authentication...\n');
      console.log('   If your browser doesn\'t open, visit:');
      console.log(`   ${authUrl.toString().slice(0, 80)}...\n`);
      
      openBrowser(authUrl.toString());
      console.log('⏳ Waiting for authentication...\n');
    });
    
    server.on('error', (err) => {
      if (err.code === 'EADDRINUSE') {
        reject(new Error(`Port ${CALLBACK_PORT} is already in use.`));
      } else {
        reject(err);
      }
    });
    
    setTimeout(() => { server.close(); reject(new Error('Login timeout')); }, 5 * 60 * 1000);
  });
}

async function checkAuth() {
  try {
    const tokenData = await fs.readFile(TOKEN_FILE, 'utf8');
    const tokens = JSON.parse(tokenData);
    const expired = tokens.expires_at && Date.now() > tokens.expires_at;
    const expiresIn = Math.floor((tokens.expires_at - Date.now()) / 1000 / 60);
    
    console.log('╔══════════════════════════════════════════════════════════════╗');
    console.log('║  🔐 AC Login Status                                          ║');
    console.log('╚══════════════════════════════════════════════════════════════╝\n');
    
    if (expired) {
      console.log('⚠️  Token expired - run `ac-login` to refresh\n');
    } else {
      console.log('✅ Logged in\n');
    }
    
    const displayName = tokens.user?.handle 
      ? `@${tokens.user.handle}` 
      : tokens.user?.email || tokens.user?.name || 'unknown';
    
    console.log(`👤 User: ${displayName}`);
    console.log(`📧 Email: ${tokens.user?.email || 'n/a'}`);
    console.log(`⏰ Expires: ${expired ? 'EXPIRED' : `in ${expiresIn} minutes`}`);
    console.log(`📁 Token: ${TOKEN_FILE}\n`);
    
    return expired ? null : tokens;
  } catch (err) {
    console.log('❌ Not logged in\n');
    console.log('   Run `ac-login` to authenticate.\n');
    return null;
  }
}

export async function getToken() {
  try {
    const tokenData = await fs.readFile(TOKEN_FILE, 'utf8');
    const tokens = JSON.parse(tokenData);
    if (tokens.expires_at && Date.now() > tokens.expires_at) return null;
    return tokens;
  } catch (err) { return null; }
}

(async () => {
  const command = process.argv[2];
  
  if (command === 'logout') {
    try { await fs.unlink(TOKEN_FILE); console.log('✅ Logged out\n'); }
    catch { console.log('ℹ️  Already logged out\n'); }
    return;
  }
  
  if (command === 'status') { await checkAuth(); return; }
  
  if (command === 'token') {
    try {
      const tokenData = await fs.readFile(TOKEN_FILE, 'utf8');
      const tokens = JSON.parse(tokenData);
      if (tokens.expires_at && Date.now() > tokens.expires_at) {
        console.error('Token expired'); process.exit(1);
      }
      console.log(tokens.access_token);
    } catch { console.error('Not logged in'); process.exit(1); }
    return;
  }
  
  if (command === 'help' || command === '-h' || command === '--help') {
    console.log(`
🔐 ac-login - Aesthetic Computer CLI Authentication

Usage:
  ac-login          Login (opens browser)
  ac-login status   Check login status  
  ac-login logout   Clear stored token
  ac-login token    Print access token (for scripts)
  ac-login help     Show this help
`);
    return;
  }
  
  await login();
})();

File: juke-wizard/Sources/JukeWizard/ACLogin.swift
// ACLogin.swift — native, in-app AC sign-in for the macOS app suite.
//
// SOURCE OF TRUTH: shared/swift/ACLogin.swift. Like ACSession.swift, the
// standalone SwiftPM apps (date-wizard, wave-wizard, clip-wizard, juke-wizard,
// slab menubar, menuband) each keep a COPY in their own Sources/. When you
// change this file, re-copy it into each consumer:
//   cp shared/swift/ACLogin.swift <app>/Sources/<App>/
//
// This is a native port of `tezos/ac-login.mjs` — the same OAuth 2.0
// Authorization-Code + PKCE flow against Auth0 (`hi.aesthetic.computer`) with a
// localhost loopback callback on port 44233. It writes the SAME ~/.ac-token
// JSON shape that ACSession.swift reads, so a sign-in here is indistinguishable
// from one done by the `ac-login` CLI — the whole suite picks it up live via
// the shared file-watch. No Terminal, no Node, no CLI dependency.
//
//   ACLogin.shared.signIn { result in
//       switch result {
//       case .success(let handle): …   // "@handle" (or email/name fallback)
//       case .failure(let error):  …
//       }
//   }
//
// The callback fires on the main queue. ACSession's file-watch will ALSO fire
// independently the instant the token lands, so UIs that already observe the
// session refresh on their own — the completion handler is for surfacing
// progress/errors on the sign-in screen itself.
import Foundation
import AppKit
import CryptoKit

final class ACLogin {
    static let shared = ACLogin()

    // ── config (must match ac-login.mjs / Auth0 allowed callback) ────────
    private let authDomain = "hi.aesthetic.computer"
    private let clientID = "LVdZaMbyXctkGfZDnpzDATB5nR0ZhmMt"
    private let callbackPort: UInt16 = 44233
    private var redirectURI: String { "http://localhost:\(callbackPort)/callback" }

    enum LoginError: LocalizedError {
        case portInUse
        case serverFailed(String)
        case stateMismatch
        case tokenExchange(String)
        case cancelled
        case timeout

        var errorDescription: String? {
            switch self {
            case .portInUse:           return "Port \(44233) is in use — close any running ac-login and retry."
            case .serverFailed(let m): return "Sign-in server failed: \(m)"
            case .stateMismatch:       return "Sign-in state mismatch — please retry."
            case .tokenExchange(let m): return "Token exchange failed: \(m)"
            case .cancelled:           return "Sign-in cancelled."
            case .timeout:             return "Sign-in timed out."
            }
        }
    }

    private var listener: SocketListener?
    private var inFlight = false

    /// True while a sign-in is waiting on the browser/callback.
    var isSigningIn: Bool { inFlight }

    // ── PKCE ─────────────────────────────────────────────────────────────
    private func base64URL(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
    private func randomURLToken(byteCount: Int) -> String {
        var bytes = [UInt8](repeating: 0, count: byteCount)
        _ = SecRandomCopyBytes(kSecRandomDefault, byteCount, &bytes)
        return base64URL(Data(bytes))
    }

    // ── public entry ───────────────────────────────────────────────────
    /// Start the browser sign-in flow. `forcePrompt` adds `prompt=login` to
    /// force the Auth0 account chooser (the CLI's `ac-login fresh`).
    func signIn(forcePrompt: Bool = false,
                completion: @escaping (Result<String, Error>) -> Void) {
        if inFlight { listener?.stop(); inFlight = false }

        let verifier = randomURLToken(byteCount: 32)
        let challenge = base64URL(Data(SHA256.hash(data: Data(verifier.utf8))))
        let state = randomURLToken(byteCount: 16)

        let finish: (Result<String, Error>) -> Void = { [weak self] result in
            guard let self else { return }
            self.inFlight = false
            self.listener?.stop()
            self.listener = nil
            DispatchQueue.main.async { completion(result) }
        }

        let server: SocketListener
        do {
            server = try SocketListener(port: callbackPort)
        } catch {
            finish(.failure(LoginError.portInUse))
            return
        }
        listener = server
        inFlight = true

        // 5-minute safety timeout, matching the CLI.
        DispatchQueue.global().asyncAfter(deadline: .now() + 300) { [weak self] in
            guard let self, self.inFlight else { return }
            finish(.failure(LoginError.timeout))
        }

        server.onRequest = { [weak self] req, respond in
            guard let self else { return }
            guard req.path.contains("callback") else {
                respond(404, "text/plain", "Not found"); return
            }
            if let err = req.query["error"] {
                let desc = req.query["error_description"] ?? ""
                respond(200, "text/html; charset=utf-8", Self.failureHTML(err, desc))
                finish(.failure(LoginError.tokenExchange(desc.isEmpty ? err : desc)))
                return
            }
            guard let code = req.query["code"] else {
                respond(400, "text/plain", "Missing authorization code"); return
            }
            guard req.query["state"] == state else {
                respond(400, "text/plain", "State mismatch")
                finish(.failure(LoginError.stateMismatch))
                return
            }
            // Exchange code → tokens, fetch userinfo + handle, write ~/.ac-token.
            self.exchangeAndStore(code: code, verifier: verifier) { result in
                switch result {
                case .success(let display):
                    respond(200, "text/html; charset=utf-8", Self.successHTML(display))
                    finish(.success(display))
                case .failure(let error):
                    respond(500, "text/plain", "Error: \(error.localizedDescription)")
                    finish(.failure(error))
                }
            }
        }

        do {
            try server.start()
        } catch {
            finish(.failure(LoginError.serverFailed("\(error)")))
            return
        }

        openBrowser(authURL(state: state, challenge: challenge, forcePrompt: forcePrompt))
    }

    /// Stop a pending sign-in (e.g. user navigated away).
    func cancel() {
        guard inFlight else { return }
        inFlight = false
        listener?.stop()
        listener = nil
    }

    // ── token exchange + persistence ─────────────────────────────────────
    private func exchangeAndStore(code: String, verifier: String,
                                  completion: @escaping (Result<String, Error>) -> Void) {
        var req = URLRequest(url: URL(string: "https://\(authDomain)/oauth/token")!)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        let body: [String: String] = [
            "grant_type": "authorization_code",
            "client_id": clientID,
            "code_verifier": verifier,
            "code": code,
            "redirect_uri": redirectURI,
        ]
        req.httpBody = try? JSONSerialization.data(withJSONObject: body)

        URLSession.shared.dataTask(with: req) { [weak self] data, resp, err in
            guard let self else { return }
            if let err { completion(.failure(err)); return }
            guard let data,
                  let http = resp as? HTTPURLResponse, http.statusCode == 200,
                  let tokens = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let access = tokens["access_token"] as? String else {
                let txt = data.flatMap { String(data: $0, encoding: .utf8) } ?? "unknown"
                completion(.failure(LoginError.tokenExchange(txt)))
                return
            }
            self.fetchUserAndWrite(tokens: tokens, access: access, completion: completion)
        }.resume()
    }

    private func fetchUserAndWrite(tokens: [String: Any], access: String,
                                   completion: @escaping (Result<String, Error>) -> Void) {
        var req = URLRequest(url: URL(string: "https://\(authDomain)/userinfo")!)
        req.setValue("Bearer \(access)", forHTTPHeaderField: "Authorization")
        URLSession.shared.dataTask(with: req) { [weak self] data, _, _ in
            guard let self else { return }
            let user = (data.flatMap {
                (try? JSONSerialization.jsonObject(with: $0)) as? [String: Any]
            }) ?? [:]
            let sub = user["sub"] as? String
            self.fetchHandle(sub: sub) { handle in
                self.writeToken(tokens: tokens, user: user, handle: handle)
                ACSession.shared.broadcastChanged()
                let display = handle.map { "@\($0)" }
                    ?? (user["email"] as? String)
                    ?? (user["name"] as? String)
                    ?? "signed in"
                completion(.success(display))
            }
        }.resume()
    }

    private func fetchHandle(sub: String?, completion: @escaping (String?) -> Void) {
        guard let sub, let encoded = sub.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed),
              let url = URL(string: "https://aesthetic.computer/handle?for=\(encoded)") else {
            completion(nil); return
        }
        URLSession.shared.dataTask(with: url) { data, _, _ in
            let handle = data.flatMap {
                (try? JSONSerialization.jsonObject(with: $0) as? [String: Any])?["handle"] as? String
            }
            completion(handle)
        }.resume()
    }

    private func writeToken(tokens: [String: Any], user: [String: Any], handle: String?) {
        let expiresIn = (tokens["expires_in"] as? Double)
            ?? (tokens["expires_in"] as? Int).map(Double.init) ?? 0
        var payload: [String: Any] = [
            "access_token": tokens["access_token"] as? String ?? "",
            "expires_at": Date().timeIntervalSince1970 * 1000 + expiresIn * 1000,
        ]
        if let r = tokens["refresh_token"] as? String { payload["refresh_token"] = r }
        if let i = tokens["id_token"] as? String { payload["id_token"] = i }
        var u: [String: Any] = [:]
        if let v = user["email"] as? String { u["email"] = v }
        if let v = user["name"] as? String { u["name"] = v }
        if let v = user["sub"] as? String { u["sub"] = v }
        if let v = user["picture"] as? String { u["picture"] = v }
        if let handle { u["handle"] = handle }
        payload["user"] = u

        guard let data = try? JSONSerialization.data(
            withJSONObject: payload, options: [.prettyPrinted]) else { return }
        // Atomic write so ACSession's file-watch sees a single clean change.
        try? data.write(to: ACSession.tokenURL, options: [.atomic])
    }

    // ── URLs / browser ────────────────────────────────────────────────
    private func authURL(state: String, challenge: String, forcePrompt: Bool) -> URL {
        var c = URLComponents(string: "https://\(authDomain)/authorize")!
        var items = [
            URLQueryItem(name: "response_type", value: "code"),
            URLQueryItem(name: "client_id", value: clientID),
            URLQueryItem(name: "redirect_uri", value: redirectURI),
            URLQueryItem(name: "scope", value: "openid profile email offline_access"),
            URLQueryItem(name: "state", value: state),
            URLQueryItem(name: "code_challenge", value: challenge),
            URLQueryItem(name: "code_challenge_method", value: "S256"),
        ]
        if forcePrompt { items.append(URLQueryItem(name: "prompt", value: "login")) }
        c.queryItems = items
        return c.url!
    }

    private func openBrowser(_ url: URL) {
        DispatchQueue.main.async { NSWorkspace.shared.open(url) }
    }

    // ── browser-facing HTML (mirrors ac-login.mjs styling) ───────────────
    private static func successHTML(_ display: String) -> String {
        """
        <!DOCTYPE html><html lang="en"><head><meta charset="UTF-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>Logged In · Aesthetic Computer</title><style>
        @media (prefers-color-scheme: dark){body{background:rgb(50,30,60);color:rgb(220,30,100)}strong{color:white}code{color:pink}}
        @media (prefers-color-scheme: light){body{background:rgb(255,240,245);color:rgb(180,20,80)}strong{color:rgb(100,20,60)}code{color:rgb(150,50,100)}}
        body{font-family:monospace;display:flex;width:100vw;height:100vh;margin:0}
        #wrapper{margin:auto;text-align:center;padding:2em}h1{font-family:sans-serif;font-weight:normal}code{font-size:120%;opacity:.75}
        </style></head><body><div id="wrapper"><h1>✅ Login Successful</h1>
        <p>Welcome, <strong>\(display)</strong>!</p><p><code>You may close this page.</code></p></div>
        <script>setTimeout(()=>window.close(),3000)</script></body></html>
        """
    }
    private static func failureHTML(_ error: String, _ desc: String) -> String {
        """
        <!DOCTYPE html><html lang="en"><head><meta charset="UTF-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>Login Failed · Aesthetic Computer</title><style>
        @media (prefers-color-scheme: dark){body{background:rgb(50,30,60);color:rgb(220,30,100)}mark{background:maroon;color:pink}code{color:pink}}
        @media (prefers-color-scheme: light){body{background:rgb(255,240,245);color:rgb(180,20,80)}mark{background:rgb(255,200,220);color:rgb(150,20,60)}code{color:rgb(150,50,100)}}
        body{font-family:monospace;display:flex;width:100vw;height:100vh;margin:0}
        #wrapper{margin:auto;text-align:center;padding:2em}h1{font-family:sans-serif;font-weight:normal}code{font-size:120%;opacity:.75}
        </style></head><body><div id="wrapper"><h1>❌ Authentication Failed</h1>
        <p><mark>\(error)</mark></p><p>\(desc)</p><p><code>You may close this page.</code></p></div></body></html>
        """
    }
}

// ── minimal loopback HTTP server (BSD sockets) ───────────────────────────
// A single-connection-at-a-time HTTP/1.0 responder. Enough to receive Auth0's
// redirect GET on 127.0.0.1 and reply with a close-this-page page. No external
// deps; avoids pulling in Network.framework just for one request.
private final class SocketListener {
    struct Request { let path: String; let query: [String: String] }

    var onRequest: ((Request, _ respond: @escaping (Int, String, String) -> Void) -> Void)?

    private let port: UInt16
    private var fd: Int32 = -1
    private let queue = DispatchQueue(label: "ac.login.socket")
    private var running = false

    init(port: UInt16) throws {
        self.port = port
        fd = socket(AF_INET, SOCK_STREAM, 0)
        guard fd >= 0 else { throw ACLogin.LoginError.serverFailed("socket()") }
        var yes: Int32 = 1
        setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, &yes, socklen_t(MemoryLayout<Int32>.size))

        var addr = sockaddr_in()
        addr.sin_family = sa_family_t(AF_INET)
        addr.sin_port = port.bigEndian
        addr.sin_addr.s_addr = inet_addr("127.0.0.1")
        let bound = withUnsafePointer(to: &addr) {
            $0.withMemoryRebound(to: sockaddr.self, capacity: 1) {
                bind(fd, $0, socklen_t(MemoryLayout<sockaddr_in>.size))
            }
        }
        guard bound == 0 else {
            close(fd); fd = -1
            throw ACLogin.LoginError.portInUse
        }
        guard listen(fd, 4) == 0 else {
            close(fd); fd = -1
            throw ACLogin.LoginError.serverFailed("listen()")
        }
    }

    func start() throws {
        running = true
        queue.async { [weak self] in self?.acceptLoop() }
    }

    func stop() {
        running = false
        if fd >= 0 { close(fd); fd = -1 }
    }

    private func acceptLoop() {
        while running {
            let client = accept(fd, nil, nil)
            if client < 0 { break }
            handle(client)
            // One callback redirect is all we need; keep looping in case the
            // browser retries (favicon, double-fetch) until stop() closes fd.
        }
    }

    private func handle(_ client: Int32) {
        defer { close(client) }
        var buf = [UInt8](repeating: 0, count: 8192)
        let n = read(client, &buf, buf.count)
        guard n > 0 else { return }
        let raw = String(decoding: buf[0..<n], as: UTF8.self)
        guard let line = raw.split(separator: "\r\n").first else { return }
        let parts = line.split(separator: " ")
        guard parts.count >= 2 else { return }
        let target = String(parts[1])

        let comps = URLComponents(string: "http://localhost\(target)")
        let path = comps?.path ?? target
        var query: [String: String] = [:]
        for item in comps?.queryItems ?? [] { query[item.name] = item.value }

        let request = Request(path: path, query: query)
        var responded = false
        let respond: (Int, String, String) -> Void = { status, contentType, body in
            guard !responded else { return }
            responded = true
            let bytes = Array(body.utf8)
            let header = "HTTP/1.0 \(status) \(status == 200 ? "OK" : "Error")\r\n"
                + "Content-Type: \(contentType)\r\n"
                + "Content-Length: \(bytes.count)\r\n"
                + "Connection: close\r\n\r\n"
            let out = Array(header.utf8) + bytes
            _ = out.withUnsafeBytes { write(client, $0.baseAddress, $0.count) }
        }

        if let onRequest {
            onRequest(request, respond)
        } else {
            respond(404, "text/plain", "Not found")
        }
        // Give async respond() (token exchange) time to flush before close.
        if !responded {
            let deadline = Date().addingTimeInterval(310)
            while !responded && Date() < deadline && running {
                usleep(50_000)
            }
        }
    }
}
